packages/core/src/hydration/utils.ts

```diff
@@ -0,0 +1,116 @@
+
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Injector} from '../di/injector';
+import {ViewRef} from '../linker/view_ref';
+import {RElement} from '../render3/interfaces/renderer_dom';
+import {isRootView} from '../render3/interfaces/type_checks';
+import {HEADER_OFFSET, LView, TVIEW, TViewType} from '../render3/interfaces/view';
+import {makeStateKey, TransferState} from '../transfer_state';
+import {assertDefined} from '../util/assert';
+
+import {NghDom, NghDomInstance} from './interfaces';
+
+/**
+ * The name of the key used in the TransferState collection,
+ * where hydration information is located.
+ */
+const TRANSFER_STATE_TOKEN_ID = '__ɵnghData__';
+
+/**
+ * Refers to the hydration data in the transferred state.
+ */
+export const NGH_DATA_KEY = makeStateKey<Array<NghDom>>(TRANSFER_STATE_TOKEN_ID);
+
+/**
+ * The name of the attribute that would be added to host component
+ * nodes and contain a reference to a particular slot in transferred
+ * state that contains the necessary hydration info for this component.
+ */
+export const NGH_ATTR_NAME = 'ngh';
+
+/**
+ * Reference to a function that reads `ngh` attribute value from a given RNode
+ * and retrieves hydration information from the TransferState using that value
+ * as an index. Returns `null` by default, when hydration is not enabled.
+ *
+ * @param rNode Component's host element.
+ * @param injector Injector that this component has access to.
+ */
+let _retrieveNghInfoImpl: typeof retrieveNghInfoImpl = (rNode: RElement, injector: Injector) =>
+    null;
+
+export function retrieveNghInfoImpl(rNode: RElement, injector: Injector): NghDomInstance|null {
+  const nghAttrValue = (rNode as HTMLElement).getAttribute(NGH_ATTR_NAME);
+  if (nghAttrValue == null) return null;
+
+  let data: NghDom = {};
+  // An element might have an empty `ngh` attribute value (e.g. `<comp ngh="" />`),
+  // which means that no special annotations are required. Do not attempt to read
+  // from the TransferState in this case.
+  if (nghAttrValue !== '') {
+    const transferState = injector.get(TransferState, null, {optional: true});
+    if (transferState !== null) {
+      const nghData = transferState.get(NGH_DATA_KEY, []) ?? [];
```

Comment:
` ?? []` is redundant since you are providing a default value.

```suggestion
      const nghData = transferState.get(NGH_DATA_KEY, []);
```